### PR TITLE
Add syntax support for CREATE FUNCTION

### PR DESCRIPTION
**Pull request by caithagoras**
Introduce CREATE FUNCTION syntax as according to https://github.com/prestodb/presto/issues/13254
```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Comment by caithagoras** [comment]
This PR is ready for review.

**Comment by rongrong** [comment]
At least add another commit to hook this up to call `FunctionManager.addFunction` so it's complete.

File: presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java [hunk 1]
```diff
@@ -575,8 +585,8 @@ public void testPrecedenceAndAssociativity()
                         new LongLiteral("2")),
                 new LongLiteral("3")));
 
-        assertExpression("1 / 2 / 3", new ArithmeticBinaryExpression(ArithmeticBinaryExpression.Operator.DIVIDE,
-                new ArithmeticBinaryExpression(ArithmeticBinaryExpression.Operator.DIVIDE,
+        assertExpression("1 / 2 / 3", new ArithmeticBinaryExpression(DIVIDE,
```
Comment (rongrong) [inline presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java]: Unrelated change.

File: presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java [hunk 2]
```diff
@@ -1413,6 +1423,64 @@ public void testCreateView()
...
     }
 
+    @Test
+    public void testCreateFunction()
+    {
+        assertStatement(
+                "CREATE FUNCTION tan (x double)\n" +
+                        "RETURNS double\n" +
+                        "SQL\n" +
+                        "DETERMINISTIC\n" +
```
Comment (rongrong) [inline presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java]: Add test for default `NOT DETERMINISTIC` and `RETURNS NULL ON NULL INPUT`.

File: presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4 [hunk 3]
```diff
@@ -139,6 +144,39 @@ property
...
+
+routineCharacteristic
+    : language
+    | determinism
+    | sqlDataAccess
+    | nullCallClause
+    ;
+
+language
+    : SQL
```
Comment (rongrong) [inline presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4]: `LANGUAGE` is a keyword. It should be `LANGUAGE language_name`

File: presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4 [hunk 4]
```diff
@@ -139,6 +144,39 @@ property
...
+    ;
+
+determinism
+    : DETERMINISTIC
+    | NOT DETERMINISTIC;
+
+sqlDataAccess
```
Comment (rongrong) [inline presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4]: I'm inclined to not include this for now because there's only one option and there's no use of this data. It's always safe to assume when language is "SQL" it will be "contains sql" anyways.

Comment (caithagoras) [reply presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4]: `SqlDataAccess` can also be `READS SQL DATA` and `MODIFIES SQL DATA` when language is `SQL`. Since this is trivial to add, we should just do it.

Syntax rules only dictates how the query text is structured and is not a bad thing to provide the flexibility to use to include or exclude this clause.

It's also safe to assume that the language is `SQL` when the body is a return statement, but we still are inclining to introducing the language clause.

File: presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java (hunk 2 above)
Comment (caithagoras) [reply presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java]: Line 1464-1468 covers those cases.

File: presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4 (hunk 4 above)
Comment (rongrong) [reply presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4]: Pretty much all other databases providing SQL function support do not require this, which is partly why I think it's probably not necessary. At this point I think it's ok if you want to introduce this but make it optional, same as `LANGUAGE SQL` is optional. Though SQL spec says it should be specified, I don't see a point. And unlike other SQL statement, `CREATE FUNCTION` doesn't have a strict backward compatibility issue if you ever need to require this to be specified.

**Comment by caithagoras** [comment]
@rongrong Removed `SqlDataAccess`. Looking for a committer approval. Thanks!

File: presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4 [hunk 5]
```diff
@@ -55,6 +55,11 @@ statement
     | ANALYZE qualifiedName (WITH properties)?                         #analyze
     | CREATE (OR REPLACE)? VIEW qualifiedName AS query                 #createView
     | DROP VIEW (IF EXISTS)? qualifiedName                             #dropView
+    | CREATE (OR REPLACE)? FUNCTION functionName=qualifiedName
+      '(' (parameterDeclaration (',' parameterDeclaration)*)? ')'
+      RETURNS returnType=type
+      routineCharacteristics
+      RETURN body=expression                                           #createFunction
```
Comment (rongrong) [inline presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4]: Let's make this future proof. `RETURN should be part of the body`. `RETURN expression` is one type of function body. Otherwise there will be major structural changes when we introduce other languages, or more complex SQL functions.

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/ParameterDeclaration.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.tree;
+
+import java.util.Objects;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static java.util.Objects.requireNonNull;
+
+public final class ParameterDeclaration
```
Comment (rongrong) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/ParameterDeclaration.java]: I don't know whether there could be other type of "parameter declaration"s. So maybe let's stick to `SqlParameterDeclaration` as in SQL Spec?

**Comment by caithagoras** [comment]
comments addressed

File: presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4 (hunk 3 above)
Comment (caithagoras) [reply presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4]: Updated `routineCharacteristic` rule

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/RoutineCharacteristics.java [hunk 7]
```diff
@@ -0,0 +1,124 @@
+/*
```
Comment (kaikalur) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/RoutineCharacteristics.java]: A general comment, not specific to this - for things like this, it may be simpler to just add these as attributes to the CreateFunction node to reduce code bloat and make it easy to work with the visitor.

Comment (caithagoras) [reply presto-parser/src/main/java/com/facebook/presto/sql/tree/RoutineCharacteristics.java]: This will be a reusable unit in `ALTER FUNCTION` so it makes sense to have it separated. Visitor won't touch this class since it is not a subclass of `Node`, or a child of `Create Function`.

File: presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4 [hunk 8]
```diff
@@ -472,21 +507,21 @@ nonReserved
     // IMPORTANT: this rule must only contain tokens. Nested rules are not supported. See SqlParser.exitNonReserved
     : ADD | ADMIN | ALL | ANALYZE | ANY | ARRAY | ASC | AT
     | BERNOULLI
-    | CALL | CASCADE | CATALOGS | COLUMN | COLUMNS | COMMENT | COMMIT | COMMITTED | CURRENT | CURRENT_ROLE
-    | DATA | DATE | DAY | DESC | DISTRIBUTED
+    | CALL | CALLED | CASCADE | CATALOGS | COLUMN | COLUMNS | COMMENT | COMMIT | COMMITTED | CURRENT | CURRENT_ROLE
```
Comment (arhimondr) [inline presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4]: question: none of the newly introduced keywords are reserved? Did we check the spec?

File: presto-parser/src/main/java/com/facebook/presto/sql/SqlFormatter.java [hunk 9]
```diff
@@ -852,6 +873,34 @@ private String formatPropertiesSingleLine(List<Property> properties)
...
+            if (sqlParameterDeclarations.isEmpty()) {
+                return "()";
+            }
+            String parameterList = sqlParameterDeclarations.stream()
+                    .map(parameter -> format(
+                            "%s%s %s",
+                            INDENT,
+                            formatExpression(parameter.getName(), parameters),
+                            parameter.getType()))
+                    .collect(joining(",\n"));
```
Comment (arhimondr) [inline presto-parser/src/main/java/com/facebook/presto/sql/SqlFormatter.java]: nit: `joining` can accept prefix / suffix that in this case is `(` , `)`

File: presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java [hunk 10]
```diff
@@ -2161,6 +2192,48 @@ private String getType(SqlBaseParser.TypeContext type)
...
+            if (characteristic.language() != null) {
+                checkArgument(characteristic.language().SQL() != null, "Unsupported language: %s", characteristic.language().getText());
+                if (language != null) {
+                    throw new ParsingException(format("Duplicate language clause: %s", characteristic.language().getText()), getLocation(characteristic.language()));
+                }
+                language = Language.SQL;
+            }
+            else if (characteristic.determinism() != null) {
+                if (determinism != null) {
+                    throw new ParsingException(format("Duplicate deterministic characteristics: %s", characteristic.determinism().getText()), getLocation(characteristic.determinism()));
```
Comment (arhimondr) [inline presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java]: nit: `Duplicate determinism characteristics`?

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/RoutineCharacteristics.java [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+    {
+        RETURNS_NULL_ON_NULL_INPUT,
+        CALLED_ON_NULL_INPUT;
+    }
+
+    private final Language language;
+    private final Determinism determinism;
+    private final NullCallClause nullCallClause;
+
+    public RoutineCharacteristics(
```
Comment (arhimondr) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/RoutineCharacteristics.java]: This constructor is unused. Also the semantics are not obvious. Let's remove it.

Comment (rongrong) [reply presto-parser/src/main/java/com/facebook/presto/sql/tree/RoutineCharacteristics.java]: This is called in `AstBuilder` I think. It's used to provide default values when they are missing.

**Comment by rongrong** [comment]
@caithagoras There's still a test failure (error message change)
